overlord/state/change.go

```diff
@@ -282,32 +291,119 @@ func init() {
 	}
 }
 
+func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task, deps []*Task) bool {
+	taskID := t.ID()
+	computeStatus := visited[taskID]
+	switch computeStatus {
+	case taskWaitStatusComputing:
+		// Cyclic dependency detected, return false to short-circuit.
+		return false
+	case taskWaitStatusWaiting, taskWaitStatusNotWaiting:
+		return computeStatus == taskWaitStatusWaiting
+	}
+	visited[taskID] = taskWaitStatusComputing
+
+	var isWaiting bool
+depscheck:
+	for _, wt := range deps {
+		switch wt.Status() {
+		case WaitStatus:
+			isWaiting = true
+		// For 'Do' and 'Undo' we have to check whether the task is waiting
+		// for any dependencies. The logic is the same, but the set of tasks
+		// varies.
+		case DoStatus:
+			isWaiting = c.isTaskWaiting(visited, wt, wt.WaitTasks())
+			if !isWaiting {
+				// Cancel early if we detect something is runnable.
+				break depscheck
+			}
+		case UndoStatus:
+			isWaiting = c.isTaskWaiting(visited, wt, wt.HaltTasks())
+			if !isWaiting {
+				// Cancel early if we detect something is runnable.
+				break depscheck
+			}
+		}
+	}
+	if isWaiting {
+		visited[taskID] = taskWaitStatusWaiting
+	} else {
+		visited[taskID] = taskWaitStatusNotWaiting
+	}
```

Comment:
Minor suggestion to simplify the control flow: we could move this `if else` into a `defer` and then just return `false` directly if a task isn't waiting. Then the breaks wouldn't be needed. Something like:
```
diff --git a/overlord/state/change.go b/overlord/state/change.go
index f6ab0610bd..019c230d05 100644
--- a/overlord/state/change.go
+++ b/overlord/state/change.go
@@ -291,7 +291,7 @@ func init() {
 	}
 }
 
-func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task, deps []*Task) bool {
+func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task, deps []*Task) (waiting bool) {
 	taskID := t.ID()
 	computeStatus := visited[taskID]
 	switch computeStatus {
@@ -303,8 +303,15 @@ func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task
 	}
 	visited[taskID] = taskWaitStatusComputing
 
+	defer func() {
+		if waiting {
+			visited[taskID] = taskWaitStatusWaiting
+		} else {
+			visited[taskID] = taskWaitStatusNotWaiting
+		}
+	}()
+
 	var isWaiting bool
-depscheck:
 	for _, wt := range deps {
 		switch wt.Status() {
 		case WaitStatus:
@@ -315,22 +322,17 @@ depscheck:
 		case DoStatus:
 			isWaiting = c.isTaskWaiting(visited, wt, wt.WaitTasks())
 			if !isWaiting {
-				// Cancel early if we detect something is runnable.
-				break depscheck
+				return false
 			}
+
 		case UndoStatus:
 			isWaiting = c.isTaskWaiting(visited, wt, wt.HaltTasks())
 			if !isWaiting {
-				// Cancel early if we detect something is runnable.
-				break depscheck
+				return false
 			}
 		}
 	}
-	if isWaiting {
-		visited[taskID] = taskWaitStatusWaiting
-	} else {
-		visited[taskID] = taskWaitStatusNotWaiting
-	}
+
 	return isWaiting
 }
 ```

Reply:
I think this example was really clever by using defer and then multiple exit points - but I'm not entirely convinced since I don't feel it improves the readability and saves only a couple of lines, but I really love the suggestion!